File: ruby/034-tournament/tournament.rb
class Tournament

  @header = <<~HEADER
    Team                           | MP |  W |  D |  L |  P
  HEADER

  def self.calculate_point(result, reverse: false)
    return 3 if (result == 'win' && !reverse) || (result == 'loss' && reverse)
    return 1 if result == 'draw'

    0 if (result == 'loss' && !reverse) || (result == 'win' && reverse)
  end

  def self.tally(input)
    input_data = input.split("\n").map { |item| item.split(';') }
    teams = {}
    input_data.each do |data|
      team_a_name = data[0]
      team_b_name = data[1]
      result = data[2]

      teams[team_a_name] ||= {
        name: team_a_name,
        mp: 0,
        w: 0,
        d: 0,
        l: 0,
        p: 0
      }

      team_a = teams[team_a_name]
      team_a[:mp] += 1
      team_a[:w] += 1 if result == 'win'
      team_a[:d] += 1 if result == 'draw'
      team_a[:l] += 1 if result == 'loss'
      team_a[:p] += self.calculate_point(result)

      teams[team_b_name] ||= {
        name: team_b_name,
        mp: 0,
        w: 0,
        d: 0,
        l: 0,
        p: 0
      }

      team_b = teams[team_b_name]
      team_b[:mp] += 1
      team_b[:w] += 1 if result == 'loss'
      team_b[:d] += 1 if result == 'draw'
      team_b[:l] += 1 if result == 'win'
      team_b[:p] += self.calculate_point(result, reverse: true)
    end

    sorted_teams = teams.values.sort_by { |team| [-team[:p], team[:name]] }
    result = @header
    sorted_teams.each do |team|
      table = <<~TABLE
        #{team[:name].ljust(30)} | #{team[:mp].to_s.rjust(2)} | #{team[:w].to_s.rjust(2)} | #{team[:d].to_s.rjust(2)} | #{team[:l].to_s.rjust(2)} | #{team[:p].to_s.rjust(2)}
      TABLE
      result += table
    end

    result
  end
end

File: go/082-tournament/tournament.go
package tournament

import (
	"bufio"
	"errors"
	"fmt"
	"io"
	"sort"
	"strings"
)

type TeamStats struct {
	MP, W, D, L, P int
}

type Team struct {
	Name string
	*TeamStats
}

func Tally(reader io.Reader, writer io.Writer) error {
	teams := make(map[string]*TeamStats)
	scanner := bufio.NewScanner(reader)

	for scanner.Scan() {
		line := scanner.Text()
		matches := strings.Split(line, ";")

		if line == "" || string(line[0]) == "#" {
			continue
		}

		if len(matches) != 3 {
			return errors.New("invalid input")
		}

		if teams[matches[0]] == nil {
			teams[matches[0]] = &TeamStats{}
		}
		if teams[matches[1]] == nil {
			teams[matches[1]] = &TeamStats{}
		}

		teams[matches[0]].MP++
		teams[matches[1]].MP++
		result := matches[2]
		if result == "win" {
			teams[matches[0]].W++
			teams[matches[0]].P += 3
			teams[matches[1]].L++
		} else if result == "loss" {
			teams[matches[1]].W++
			teams[matches[1]].P += 3
			teams[matches[0]].L++
		} else if result == "draw" {
			teams[matches[0]].D++
			teams[matches[0]].P++
			teams[matches[1]].D++
			teams[matches[1]].P++
		} else {
			return errors.New("invalid input")
		}
	}

	var sortedTeams []Team
	for name, stats := range teams {
		sortedTeams = append(sortedTeams, Team{name, stats})
	}

	sort.Slice(
		sortedTeams, func(i, j int) bool {
			if sortedTeams[i].P == sortedTeams[j].P {
				return sortedTeams[i].Name < sortedTeams[j].Name
			}
			return sortedTeams[i].P > sortedTeams[j].P
		},
	)

	_, err := fmt.Fprintf(writer, "Team                           | MP |  W |  D |  L |  P\n")
	for _, team := range sortedTeams {
		_, err = fmt.Fprintf(
			writer, "%-30s | %2d | %2d | %2d | %2d | %2d\n",
			team.Name, team.MP, team.W, team.D, team.L, team.P,
		)
	}

	if err != nil {
		return err
	}

	return nil
}
